feat(asgi,fastapi,starlette)!: provide both send and receive hooks with `scope` and `message`

eigenein opened the pull request:
# Description

Currently, a «receive» hook is only provided with `scope`, but not `message`. On contrary, a «send» hook is only provided with `message`, but not `scope`.

This change unifies the both hooks and provides them with additional request info.

This is a breaking change:

- The hooks will now be provided with one more positional argument
- `client_request_hook` will be called _after_ `receive()` instead of _before_ `receive()`

Resolves #2560.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Existing unit tests have been updated.

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

eigenein commented:
Fixed the linter errors, let's try it again

eigenein commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -227,10 +232,6 @@ def client_response_hook(span: Span, message: dict):
 
-_ServerRequestHookT = typing.Optional[typing.Callable[[Span, dict], None]]
-_ClientRequestHookT = typing.Optional[typing.Callable[[Span, dict], None]]
-_ClientResponseHookT = typing.Optional[typing.Callable[[Span, dict], None]]
```
Extracted these into `types.py` and made public, so that they could be reused in the FastAPI and Starlette instrumentations

eigenein commented:
Ready for review

tammy-baylis-swi commented:
Thank you for submitting this! A couple of asks:

> client_request_hook will be called after receive() instead of before receive()

(1) How has this been tested? Does the hook still work the same way to e.g. set span name or attributes?

(2) Please could you also create and link a github Issue describing the feature, like how [this PR](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2538) links [this issue](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1742), especially because this would be a breaking change.

eigenein commented:
Sure!

1. There are existing tests for hooks, that verify the hooks keep working the same way (they test exactly setting span name & attributes and would've got red should I break anything):
    - [ASGI](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py#L682-L713)
    - [FastAPI](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-fastapi/tests/test_fastapi_instrumentation.py#L353-L383)
    - [Starlette](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-starlette/tests/test_starlette_instrumentation.py#L270-L304)

    Of course, they went red because I added the new parameters, so I just updated the tests.

    The [call order change here](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L670-L672) basically ensures that the hook could access the ASGI event. No functional changes, otherwise. I don't think the call order was part of the public interface, but pointed it out – just in case.

2. Would you like me to just copy-paste the description, or would need some additional info in there? This pull request is pretty self-contained, there was no bug or feature request in the first place.

tammy-baylis-swi commented:
>    The [call order change here](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L670-L672) basically ensures that the hook could access the ASGI event. No functional changes, otherwise. I don't think the call order was part of the public interface, but pointed it out – just in case.

Great thanks!

> 2. Would you like me to just copy-paste the description, or would you need some additional info in there? This pull request is pretty self-contained, there was no bug or feature request in the first place.

Yes please a copy-paste works 🙂

tammy-baylis-swi reviewed (approved):
Also Lgtm, nice to see those types in shared ASGI module. We'll get eyes from the maintainers as well.

eigenein commented:
Done!

lzchen commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -666,9 +667,9 @@ async def otel_receive():
-                    self.client_request_hook(receive_span, scope)
                 message = await receive()
+                if callable(self.client_request_hook):
+                    self.client_request_hook(receive_span, scope, message)
```
Im confused as to why we are changing the order of this? Conceptually this was supposed to represent a hook that users can add that will be called BEFORE the receive call is made and right AFTER a span is created, such as in [requests](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/__init__.py#L216) and [flask](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L437). It also follows the precedence of the [server request hook](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L617). This way users can implement logic like timing how long a request has taken.

eigenein replied:
> Conceptually this was supposed to represent a hook that users can add that will be called BEFORE the receive call is made and right AFTER a span is created, such as in [requests](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/__init__.py#L216) and [flask](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L437).

I believe that my change actually makes it more aligned with those:

- In Flask, request is _already read_ and available in [`flask.request`](https://flask.palletsprojects.com/en/3.0.x/api/#flask.request) by the moment the hook is called. Similarly, `client_request_hook` would get called right after the event has been read, but still _before it gets processed_ in the [`self.app(...)`](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L632) call down the chain
- In `requests`, the `request` is _already prepared_ by the moment the request hook is called
- In Django, the request hook is called [at the very end](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware/otel_middleware.py#L289), when `request` is fully prepared

> It also follows the precedence of the [server request hook](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L617). This way users can implement logic like timing how long a request has taken.

Server request hook gets called before [`self.app(...)`](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L632) and hence, before `otel_receive` even gets called and before the hook/receive execution order gets to play its role. So, my change does not affect it at all, and I'm not sure if I understand your point well. Could you elaborate on that?

lzchen commented on CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#2425](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2425))
 - `opentelemetry-instrumentation-flask` Add `http.method` to `span.name`
   ([#2454](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2454))
+- ASGI, FastAPI, Starlette: provide both send and receive hooks with `scope` and `message` ([#2546](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2546))
```
```suggestion
- ASGI, FastAPI, Starlette: provide both send and receive hooks with `scope` and `message` for internal spans ([#2546](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2546))
```

lzchen commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -81,15 +81,15 @@ async def hello():
             span.set_attribute("custom_user_attribute_from_request_hook", "some-value")
 
-    def client_request_hook(span: Span, scope: dict):
+    def client_request_hook(span: Span, scope: dict[str, Any], message: dict[str, Any]):
```
> The [call order change here](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L670-L672) basically ensures that the hook could access the ASGI event.

If request hooks were meant to be called right after span creation and before requests are made, should the hook even be allowed to accept the `message` field?

eigenein replied:
I don't think there is even concensus on that, but the existing instrumentations seem to more agree with me: in Django, Requests, `urllib`, `urllib3`, and HTTPX instrumentations, the request hook is able to access the request instance. It's either this or there seems to be no concensus at all. E.g. Flask is a little weird as it receives only the `environ` part of the request instance even though the latter is available.

Also, I could argue that this:

> If request hooks were meant to be called right after span creation and before requests are made

depends heavily on definition of «before request is made». In ASGI instrumentation, we have `server_request_hook` that gets called before any reading and processing happens.

lzchen replied:
Trying to consolidate what was said in [this](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2546#discussion_r1621092953) thread and the this one. I agree we should have a concensus on what these hooks should be doing and what parameters would add the most value/functionality to customers. Currently taken from the docs: 

> This instrumentation supports request and response hooks. These are functions that get called
right after a span is created for a request and right before the span is finished for the response.

This definition doesn't seem too helpful in terms of implementation as well as functionality. If we were to define how hooks would be useful, it probably would be "a place where information pertaining to the current request/response is most present so that I can use that information to run custom logic".

If we agree on this definition, then I see the value of having as much information available for the hooks as possible pertaining to those specific steps. However, I do have one question:

Also, apologies for my inexperience with ASGI, but if we were to make both client* hooks have the `message` and `scope` arguments and we change the ordering of the `client_request_hook` AFTER the actual recieve, what would be a use case in which the `client_request_hook` must be used but the `client_response_hook` can't? What would be the purpose of having both in this case?

lzchen replied:
Great! Thanks for the explanation!
